src/ripple/app/misc/NetworkOPs.cpp

```diff
@@ -1327,38 +1306,47 @@ bool NetworkOPsImp::checkLastClosedLedger (
             ++ledgers[peerLedger].nodesUsing;
     }
 
-    auto bestVC = ledgers[closedLedger];
 
     // 3) Is there a network ledger we'd like to switch to? If so, do we have
     // it?
     bool switchLedgers = false;
+    ValidationCount bestCounts = ledgers[closedLedger];
 
     for (auto const& it: ledgers)
     {
-        JLOG(m_journal.debug()) << "L: " << it.first
-                              << " t=" << it.second.trustedValidations
-                              << ", n=" << it.second.nodesUsing;
+        uint256 const & currLedger =  it.first;
+        ValidationCount const & currCounts = it.second;
 
-        // Temporary logging to make sure tiebreaking isn't broken
-        if (it.second.trustedValidations > 0)
-            JLOG(m_journal.trace())
-                << "  TieBreakTV: " << it.first;
-        else
+        JLOG(m_journal.debug()) << "L: " << currLedger
+                              << " t=" << currCounts.trustedValidations
+                              << ", n=" << currCounts.nodesUsing;
+
+        bool const preferCurr = [&]()
         {
-            if (it.second.nodesUsing > 0)
+            // Prefer ledger with more trustedValidations
+            if (currCounts.trustedValidations > bestCounts.trustedValidations)
+                return true;
+            if (currCounts.trustedValidations < bestCounts.trustedValidations)
+                return false;
+            // If neither are trusted, prefer more nodesUsing
+            if (currCounts.trustedValidations == 0)
             {
-                JLOG(m_journal.trace())
-                    << "  TieBreakNU: " << it.first;
+                if (currCounts.nodesUsing > bestCounts.nodesUsing)
+                    return true;
+                if (currCounts.nodesUsing < bestCounts.nodesUsing)
+                    return false;
             }
-        }
+            // If tied trustedValidations (non-zero) or tied nodesUsing,
+            // prefer higher ledger hash
+            return currLedger > closedLedger;
+
+        }();
 
-        // Switch to a ledger with more support
-        // or the one with higher hash if they have the same support
-        if (it.second > bestVC ||
-            (it.second == bestVC && it.first > closedLedger))
+        // Switch to current ledger if it is preferred over best so far
+        if (preferCurr)
         {
-            bestVC = it.second;
-            closedLedger = it.first;
+            bestCounts = currCounts;
+            closedLedger = currLedger;
             switchLedgers = true;
         }
     }
```

Comment:
I believe that the code from line 1313 through here can be replaced with this:

```
auto result = std::max_element(ledgers.begin(), ledgers.end(),
        [&m_journal](auto const& best, auto const& curr)
        {
            JLOG(m_journal.debug()) <<
                "L: " << curr.first <<
                ": t=" << curr.second.trustedValidations <<
                ", n=" << curr.second.nodesUsing;

            // Prefer ledger with more trustedValidations
            if (curr.second.trustedValidations > best.second.trustedValidations)
                return true;
            if (curr.second.trustedValidations < best.second.trustedValidations)
                return false;
            // If neither are trusted, prefer more nodesUsing
            if (curr.second.trustedValidations == 0)
            {
                if (curr.second.nodesUsing > best.second.nodesUsing)
                    return true;
                if (curr.second.nodesUsing < best.second.nodesUsing)
                    return false;
            }
            // If tied trustedValidations (non-zero) or tied nodesUsing,
            // prefer higher ledger hash
            return curr.first > best.first;
        });

    assert (result != ledgers.end());
```

Note, that we are guaranteed that `ledgers` is not empty and *will* contain an entry for the initial value of `closedLedger` because of line 1294:

```
auto& ourVC = ledgers[closedLedger];
```

Determining whether we need to switch ledgers now becomes trivial: simply compare `result->first` to `closedLedger`. The `bestCounts` is no longer needed.

Reply:
I like the suggestion, although I think it won't print the debug output for the first item in `ledgers`. This area is likely to see a refactor soon so I will take this idea up then.

Reply:
Yes, you’re right, it (understandably) won’t. That could be a problem when examining logs.